atc/api/accessor/verifier.go

```diff
@@ -0,0 +1,138 @@
+package accessor
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+var (
+	ErrVerificationNoToken          = errors.New("token not provided")
+	ErrVerificationInvalidToken     = errors.New("token provided is invalid")
+	ErrVerificationFetchFailed      = errors.New("failed to fetch public key")
+	ErrVerificationInvalidSignature = errors.New("token has invalid signature")
+	ErrVerificationTokenExpired     = errors.New("token is expired")
+	ErrVerificationInvalidAudience  = errors.New("token has invalid audience")
+	ErrVerificationFailed           = errors.New("token verification failed")
+)
+
+func NewVerifier(httpClient *http.Client, keySetURL *url.URL, audience []string) *verifier {
+	return &verifier{
+		httpClient: httpClient,
+		keySetURL:  keySetURL,
+		audience:   audience,
+	}
+}
+
+type verifier struct {
+	sync.Mutex
+	httpClient  *http.Client
+	keySetURL   *url.URL
+	keySet      jose.JSONWebKeySet
```

Comment:
As an optimization, would it make sense to pass this in to the verifier as it could be cached higher up - for now even on bootstrapping as we don't expect this key to change dynamically ?

Reply:
There's only ever one `verifier`, which is created in the atccmd: https://github.com/concourse/concourse/pull/4950/files#diff-67f9436106f0941622f0345dbf1c82a2R1675-R1685

Reply:
See https://github.com/concourse/concourse/pull/4950#discussion_r372619432

Reply:
@cirocosta 's suggestion was to leverage something [once](https://golang.org/pkg/sync/#Once) to fetch lazily but only once.

Reply:
just checking for `ErrVerificationInvalidSignature`

Reply:
sounds good!